### DISCUSSION TITLE
Commenting code should respect the indentation

### DISCUSSION THREAD

**Discussion by Chris2011** [apache/netbeans#5145] [Ideas]
As the title says, commenting code should respect the indentation of the code.
There are some excetptions but I will cover this later.

Atm the current comment logic for Java and CSL languages (w/o block comment) behaves like this:

```typescript
//class test {
//  test() {
//    if(true) {
//      if(true) {
//
//      }
//    }
//  }
//}
```

it searches for the linestart and adds the comment prefix to the code. If I comment
the whole file, this is not a problem but if I just comment an inner code block, it will looks like this:

```typescript
class test {
  test() {
//    if(true) {
//      if(true) {
//
//      }
//    }
  }
}
```

the problem here bevahes after formatting the code:

```typescript
class test {
  test() {
    //    if(true) {
    //      if(true) {
    //
    //      }
    //    }
  }
}
```

The code reformats because comments should be also part of the indentation and when I remove the comments
it needs to be reformatted again. So the wanted behaviour should be this:

```typescript
class test {
  test() {
    //if(true) {
    //  if(true) {
    //
    //  }
    //}
  }
}
```

This code doesn't needs to be reformatted because it is on the same indentation as a block.
Yes there are some other cases where it still needs to reformatted but this is ok:

```typescript
class test {
  test() {
    //if(true) {
      if(true) {

      }
    //}
  }
}
```

After reformatting it looks like this:

```typescript
class test {
  test() {
    //if(true) {
    if(true) {

    }
    //}
  }
}
```

For HTML and CSS/LESS/SCSS/SASS it looks like this and behaves a bit better:

```html
<div>
  <div>
    <!--<div></div>-->
  </div>
</div>
```

```scss
body {
  /*color: green;*/
}
```

But for multiline it looks like this then:

```html
<div>
<!--  <div>
    <div></div>
  </div>-->
</div>
```

```scss
body {
/*  div {
    div {

    }
  }*/
}
```

The code for CSL is somewhere around here: https://github.com/apache/netbeans/blob/4f8fbb4bd1243d787dc60adef9b3761e791de9be/ide/csl.api/src/org/netbeans/modules/csl/api/ToggleBlockCommentAction.java#L507
For Java needs to look deeper.

**Comment by Chris2011** [comment]
I already had a look and found this function `LineDocumentUtils.getNextNonWhitespace(doc, ` and when I change the line that I mentioned, it works quite well for a single line, but logic behind that is a bit more complex, because if the next line is an inner block I'm not sure what is best then, should it behave like I "fixed" it or should it be on the same offset, where the first comment was added? I looked in VS Code and WebStorm and both are behaving a bit different but intellligent enough. So open for discussions here what should be best. We are talking about single line commenting and selected code blocks those are also treated like single lines.

**Reply by mbien** [reply]
if the formatter would ignore comments which have no spaces before `//`, this would solve most of the issues listed here, right?

**Reply by Chris2011** [reply]
No, because it should be formatted because if the indentation. As far as I can remember. There was an option for indentation inside formatting for comments: respect indentation on comments. So this needs to be fixed IMHO

**Reply by mbien** [reply]
I am not sure. If a user wants to align the comment with the indentation, then yes.
example:
```java
    public static void main(String[] args) {
        // formatted comment on same level as the code
        System.out.println("Hello There!");
    }
```

but if something is simply commented out like:
```java
    public static void main(String[] args) {
//        System.out.println("Hello There!");
    }
```
I am not sure if there would be any benefit in formatting this. Thats why I thought that that formatter could simply ignore un-indented comments while trying to format the rest.

**Reply by Chris2011** [reply]
That's why we need to make it customizable but IMHO it works exactly like in other IDEs and this seems just fine for me. If the comment is exactly where the token starts is the same where it doesn't matter whether it was just commented out for whatever reason. The reason shouldn't matter. The current behavior looks just annoying and is not just annoying it behaves in a weird way. Ignoring is just a workaround for this, yes but doesn't feel right IMHO.

**Comment by Chris2011** [comment]
Little example where the comments should always be part of the indentation, also if you see code with comments they are aligned to 99%.

```java
    private void test2() {
// Erstellen Sie eine neue Instanz von JToolBar
        JPanel toolBar2 = new JPanel();
// Setzen Sie die Größe und die Farbe der neuen Toolbar
        toolBar2.setPreferredSize(new Dimension(100, 30));
        toolBar2.setBackground(Color.LIGHT_GRAY);
        toolBar2.add(new JLabel("dasdadsasd"));

// Holen Sie sich das Hauptfenster der Anwendung
        JFrame mainWindow = (JFrame) WindowManager.getDefault().getMainWindow();

// Holen Sie sich das Container-Panel des Hauptfensters
        Container contentPane = mainWindow.getContentPane();

// Fügen Sie das toolBar2-Panel dem Container-Panel hinzu
        contentPane.add(toolBar2, BorderLayout.PAGE_START);

// Aktualisieren Sie das Fenster, um die Änderungen anzuzeigen
        JToolBar mainToolBar = (JToolBar) mainWindow.getContentPane().getComponent(1);

// Fügen Sie die Toolbar dem Container-Panel hinzu
        contentPane.add(mainToolBar, BorderLayout.PAGE_END);

// Aktualisieren Sie das Fenster, um die Änderungen anzuzeigen
        mainWindow.revalidate();
    }
```

I copied the code within the test2 method from a ChatGPT example. So everything started at linestart. After formatting, it just formatted the code and not the comments and this looks very ugly and is really not useful IMHO.
